### ISSUE TITLE
Implement Preferences

### ISSUE THREAD

**Issue by lah7** [lah7/razer_blackwidow_chroma_gui#3]
A place to change some features of the application to suit the user's tastes:

**Planned:**
* Changing keyboard layout.
* Toggle "live switching" when changing profiles.
* Toggle "live preview" while editing profiles.
* Toggle whether profiles should be activated immediately after saving.

**Possibly:**
* Develop a separate script to auto start certain effects when the user logs on, but integrate this into the preferences.

**Comment by terrycain** [comment]
> Changing keyboard layout.

Is a priority now that we have an AZERTY layout. Once the guy has got back to me with the keyboard mapping answers I'll see the best way to handle that. Then it should be as simple as adding another layout.

**Comment by lah7** [comment]
> Changing keyboard layout

The idea is a Preferences page where the user could use a dropdown to pick their keyboard layout, eg. "English (United Kingdom)" and "English (United States)".

I imagine the key positions will be relatively the same, it'll just be what that key is that needs amending.

The keyboard files aren't perfect either, I might redo that by having fixed X,Y co ords as opposited to using clumsy tables.

**Comment by terrycain** [comment]
I'm thinking of creating a SVG of the keyboard to improve the layout of the layout page

**Comment by lah7** [comment]
A SVG sounds good, we could even add transparency and create a wave effect underneath it. :smirk:

I already have my own scan and high-res crop as a PNG (no stealing internet images :wink:), it shouldn't be too hard redrawing that - my question is how will be buttons appear on the page?

**Comment by terrycain** [comment]
I started tracing the internet one as technically we not redistributing that image so it fine + the SVG will be accurate.

As long as the ratio of the keyboard is kept as it was traced it should be good. Good idea on the wave effect, didn't think of that ;)

**Comment by terrycain** [comment]
![SVG progress](http://s14.postimg.org/yoish1a7l/Screenshot_from_2016_01_14_00_41_54.png)

**Comment by lah7** [comment]
Good job. We'll probably need a couple of versions of these, like the smaller LSHIFT key and "\ |" key. I'm curious if the x,y co-ords are pretty much identical across versions.

While still a work in progress, I'd love it if Razer would let us use their font... if there's a similar open source one, even better! :wink:

**Comment by terrycain** [comment]
Have a look for an opensource one and ill see if I can include it. Don't bother emailing them, I tried once and they just replied with generic rubbish, I highly doubt they'll help us.

Yeah I was just tracing the US one, making a UK one would be a priority ;)

**Comment by terrycain** [comment]
As for the x,y coords. Yes I am too curious about that. In theory they would of been lazy in programming the keyboard so they should just work exactly the same, I had always assumed that's why I have gaps in the mapping table I made

**Comment by lah7** [comment]
Well, when it comes to the UI, the [Play](https://www.google.com/fonts/specimen/Play) font (under the [SIL Open Font License](SIL Open Font License, 1.1) - not certain if GPL-compatible?) looks the closest to RazerF5, but the keyboard font will be trickier to find.

![screenshot at 2016-01-14 14 31 34](https://cloud.githubusercontent.com/assets/13032135/12326979/cc2f733c-bacb-11e5-9dd7-d72d3b4fe9dc.png)

**Left** = Razer Regular ([Fan-made? License not explicitly stated.](http://liquidhero.deviantart.com/art/Razer-regular-font-323069609))
**Middle** = RazerF5 (their web font - all rights reserved)
**Right** = Play (via Google Fonts)

**Comment by terrycain** [comment]
I think if we actually want it to look like the keyboard we'll need to make a font. Which in theory shouldn't be **too** difficult.

**Comment by terrycain** [comment]
![screenshot from 2016-01-14 23 26 28](https://cloud.githubusercontent.com/assets/1986397/12341012/7b7415c4-bb16-11e5-9928-2cc4bc144939.png)

**Comment by terrycain** [comment]
That 1 SVG currently contains the en-gb and en-us layouts and once we decide how to choose layouts (im thinking just take the system default) it can swap between them.

**Comment by lah7** [comment]
Looking very nice, good job! Only one tiny issue -- use of Razer's logo?

**Comment by terrycain** [comment]
I'm sure that will be fine for now. Razer have far better things to be doing than watching our repo.

**Comment by lah7** [comment]
True, but with Git leaving traces, it's still a violation of the GPL. Either way, you could use the "snakes" inspired logo for these drivers, specifically intended to avoid that.

**Comment by lah7** [comment]
**A round-up on this:**

> Changing keyboard layout.

Not essential as it's detected automatically.


> Toggle "live switching" when changing profiles.

Implemented in 18afc825fa2022df32c7eab13c1a3062dfbc4a74

> Toggle "live preview" while editing profiles.
> Toggle whether profiles should be activated immediately after saving.

Also implemented in 18afc825fa2022df32c7eab13c1a3062dfbc4a74 but the "backend" code to read the profile directly from memory is not implemented yet.

`fixme:activate_profile_from_memory`

Should be simple. Feel free to jump ahead and'fix this if you desire. I haven't really looked into your new code in detail yet.

---------
> Possibly:
>
>Develop a separate script to auto start certain effects when the user logs on, but integrate this into the preferences.

Could be a future feature or part of an entirely different script.

**Comment by terrycain** [comment]
Will look into it in a bit ;).

----

> Develop a separate script to auto start certain effects when the user logs on

The tray applet already starts on log on, we could just get that to read a config file containing a `startup` section and it just actions whatever is in there?

**Comment by terrycain** [comment]
ooh looking real nice, seems to be some errors from the javascript though

**Comment by lah7** [comment]
Yeah, sadly I broke something in 45a7790d1ee0c6305dae2c3e0a04e19a12cd9d66 (details there)

To be expected, the page is loaded "fresh" every time now, whereas before everything was piled onto one page. I suspect the order of which variables are being declared need adjusting, I've been focused on other things. :wink:

**Comment by terrycain** [comment]
Separating out onto different pages is a good idea. Yeah fixing now, your calling things in the wrong order ;)

**Comment by terrycain** [comment]
+1 for using xdg-open

**Comment by lah7** [comment]
> The tray applet already starts on log on, we could just get that to read a config file containing a startup section and it just actions whatever is in there?

Sorry, I nearly overlooked this. That sounds like a plan. We've already got a `preferences.json` now, so we could use that -- just like it's going to remember the previously used colour.

**Comment by terrycain** [comment]
Found the source of the issue -_- 

Basically as javascript is super asynchronous, even if we call `keyboard_obj.load()` it still takes a little time for snapsvg to load the svg, and in that little time JavaScript will have tried to run all the commands after it. 

Ive altered the `keyboard_obj` to take a callback function, which it will call **once** snap has loaded the SVG. But we only want to call load once the page has loaded. So I've made a class which accumulates JavaScript commands and wraps them all in the jQuery onload function. The class can be used to wrap anything so I used it to wrap itself wrapping the `keyboard_obj.load()` then at the end it runs it all.

So for example the profile_editor section line 80
```

js_exec = WebkitJavaScriptExecutor(self.webkit) # Wraps document.on_load

# Wraps keyboard.load
kb_callback = WebkitJavaScriptExecutor(None, wrapper="keyboard_obj.load(function(){{{0}}});")

... all the set key commands ...

Then kb_callback would be ran inside of js_exec which would then be ran by webkit.
```
Enjoy

Fixed live preview and the preview button.

Found issue where editing and then cancelling doesn't revert changes.

**Comment by lah7** [comment]
Excellent! Thank you for looking into this.

Send it in as a pull request here, I have a few more changes to make, another test of everything, and then it's ready for upstream. :smile:

**Comment by terrycain** [comment]
I can't make a fork of your repository as github wont allow it. Mind giving me write access and ill push to a new branch

**Comment by lah7** [comment]
Done, should be good to go.

**Comment by terrycain** [comment]
Should now have a branch `18012015_gui_fixes`

**Comment by lah7** [comment]
Preferences implemented in 18afc825fa2022df32c7eab13c1a3062dfbc4a74 and all good to go with fixes from ddbab435ee72d1d1cde81f3b137f06ed8a00d8c7
